File: src/Rulesage.Graph/Extensions/DependencyInjectionExtensions.fs
namespace Rulesage.Graph.Extensions

open System.Runtime.CompilerServices
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Rulesage.Graph
open Rulesage.Graph.Services.Abstractions
open Rulesage.Graph.Services.Implementations

type ServiceCollectionExtensions =
    [<Extension>]
    static member AddGraphModule(services: IServiceCollection, config: IConfiguration) =
        services.Configure<GraphConfig>(config.GetSection("Graph")) |> ignore

        services
            .AddScoped<IStructureBuilder, StructureBuilder>()
            .AddScoped<IDescriptionCleaner, DescriptionCleaner>()
            .AddScoped<ISemanticGraphBuilder, SemanticGraphBuilder>()
            .AddScoped<IGraphFuser, GraphFuser>()
            .AddScoped<IGraphBuilder, GraphBuilder>()
            .AddScoped<ILabelPropagator, LabelPropagator>()
            .AddScoped<IGraphDotExporter, GraphDotExporter>()
        |> ignore

        services

File: src/Rulesage.Retrieval/Extensions/DependencyInjectionExtensions.cs
using Microsoft.Extensions.DependencyInjection;
using Rulesage.Retrieval.Options;

namespace Rulesage.Retrieval.Extensions;

public static class ServiceCollectionExtensions
{
    extension(IServiceCollection collection)
    {
        public IServiceCollection AddOperationRetrieval(
            Action<RetrievalOptions>? configureOptions = null)
        {
            collection.Configure(configureOptions ?? (_ => { }));

            collection.AddScoped<IRuleRetrievalService, RuleRetrievalService>();

            return collection;
        }
    }
}